components/connector/config/servers.go

```diff
@@ -33,13 +35,19 @@ func PrepareExternalGraphQLServer(cfg Config, certResolver api.CertificateResolv
 
 	externalRouter.Use(authContextMiddleware)
 
-	return &http.Server{
-		Addr:    cfg.ExternalAddress,
-		Handler: externalRouter,
+	handlerWithTimeout, err := timeouthandler.WithTimeout(externalRouter, cfg.ServerTimeout)
+	if err != nil {
+		return nil, err
 	}
+
+	return &http.Server{
+		Addr:              cfg.ExternalAddress,
+		Handler:           handlerWithTimeout,
+		ReadHeaderTimeout: cfg.ServerTimeout,
```

Comment:
im not sure if when the server read header timeout hits and the same time as the handlerWithTimeout, that you would be allowed to write the timeout handler message to the socket (this can be tested by making the ReadHeaderTimeout slightly less than the handlerWithTimeout timeout. If this is the case, we should have the ReadHeaderTimeout be equal to  cfg.ServerTimeout + 1 second

Reply:
ReadHeaderTimeout is hit before even ServeHTTP of the timeout handler is invoked. Therefore it is a timeout before the actual server timeout respected by the timeout handler is hit.

Therefore if the headers are read 10 seconds the time for the timeout handler will still be 110s and the whole request will be 120s long. But I think that this is not an issue at all because in our istio setup we only allow some specific headers that we know are safe and will never be needed so much time for them to be fetched.

Another approach could be to come up with a dedicated timeout for the headers and subtract it from the server timeout this way server processing will be readHeaderTimeout + ServerTimeout by timeout handler.